src/openassetio-python/package/openassetio/test/manager/apiComplianceSuite.py

```diff
@@ -513,6 +513,270 @@ def __testRegisterError(self, fixture_name, errorCode):
+        a_rel = self.a_traits_data_with_properties_for_a_known_relationship
+        expected = (
+            self.expected_relation_references_for_a_traits_data_with_properties_for_a_known_relationship
+        )
+        self.__testGetWithRelationships([a_rel], a_ref, [expected])
+
+    def test_when_result_trait_set_supplied_then_filtered_refs_returned(self):
+        a_ref = self.a_reference_to_a_readable_entity_with_relations
+        a_rel = TraitsData(self.a_trait_set_for_a_known_relationship)
+        result_trait_set = self.an_entity_trait_set_for_a_subset_of_relations
+        expected = (
+            self.expected_relation_subset_references_for_a_trait_set_for_a_known_relationship
+        )
+        self.__testGetWithRelationships([a_rel], a_ref, [expected], result_trait_set)
+
+    def test_when_querying_missing_reference_then_resolution_error_is_returned(self):
+        self.__testGetWithRelationshipsError("a_reference_to_a_missing_entity")
+
+    def test_when_querying_malformed_reference_then_malformed_reference_error_is_returned(self):
+        self.__testGetWithRelationshipsError(
+            "a_malformed_reference",
+            errorCode=BatchElementError.ErrorCode.kMalformedEntityReference,
+        )
+
+    def __testGetWithRelationships(
+        self, relationships, reference, expected_relations, resultTraitSet=None
+    ):
+        context = self.createTestContext(access=Context.Access.kRead)
+        results = []
+
+        self._manager.getWithRelationships(
+            relationships,
+            reference,
+            context,
+            lambda _idx, traits_data: results.append(traits_data),
+            lambda idx, batch_element_error: self.fail(
+                f"getWithRelationships should not error for index {idx}: "
+                f"{batch_element_error.message}"
+            ),
+            resultTraitSet,
+        )
+
+        self.assertEqual(results, expected_relations)
+
+    def __testGetWithRelationshipsError(
+        self,
+        fixture_name,
+        access=Context.Access.kRead,
+        errorCode=BatchElementError.ErrorCode.kEntityResolutionError,
+    ):
+        reference = self._manager.createEntityReference(
+            self.requireFixture(fixture_name, skipTestIfMissing=True)
+        )
+
+        expected_msg = self.requireFixture(f"the_error_string_for_{fixture_name}")
+        expected_error = BatchElementError(errorCode, expected_msg)
+
+        context = self.createTestContext()
+        context.access = access
+
+        a_rel = TraitsData(self.a_trait_set_for_a_known_relationship)
+
+        results = []
+        self._manager.getWithRelationships(
+            [a_rel],
+            reference,
+            context,
+            lambda _idx, _refs: self.fail("Unexpected success callback"),
+            lambda _idx, batch_element_error: results.append(batch_element_error),
+        )
+        [actual_error] = results  # pylint: disable=unbalanced-tuple-unpacking
```

Comment:
[consider] The exception raised if no callback was ever called isn't very informative
```
ValueError: not enough values to unpack (expected 1, got 0)
```
One idea: initialise to `results = ["callback not called"]`, so the error log generated by the `assertEqual` has some hint as to what happened?

Another idea: `self.assertEqual(len(results), 1, msg="Callback should have been called once")`

:shrug:

Reply:
Hmm, quick experimentation finds the output isn't great, would take some experimentation to find the nicest assertion function.  E.g. output of `assertEqual` with a custom `msg`
```python
>>> class test(unittest.TestCase):
...   def test(self):
...     self.assertEqual(["a"], [], msg="Callback should have been called once")
... 
>>> test().test()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in test
  File "/usr/lib/python3.8/unittest/case.py", line 912, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python3.8/unittest/case.py", line 1118, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/usr/lib/python3.8/unittest/case.py", line 1100, in assertSequenceEqual
    self.fail(msg)
  File "/usr/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: Lists differ: ['a'] != []

First list contains 1 additional elements.
First extra element 0:
'a'

- ['a']
+ [] : Callback should have been called once
```

Reply:
Capturing off-line discussion - keeping these things out-of-scope at this point as it also affects other existing tests.